From 435c775d4e4262b7b81e41ed1604df2ec76fc0c3 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 145/296] fix force stop home app may black screen.

Symptom: when doing force stop home app in foreground,
if launch google recent app before, will show black screen,
press back key no responce, must press home key or recent key
to re-launch home activity.

Root Cause: behindFullscreenActivity set as true caused
            resume next google recent app activity after force stop
            but quickly recent activity been added into stopped list by
            makeInvisible, then no other activity can resumed.

This case can be reproduced by android 7.1.1 emulator by below steps:
 1. press recent key to launch google recent app.
 2. press home key to back home.
 3. use "adb shell am force-stop com.google.android.apps.nexuslauncher"
 4. black screen after then.

Solution: set behindFullscreenActivity by task.getTopActivity to check
          if any top actiivty for home stack can resume.
          for aligning android 6.0 behavior to resume recent activity.
---
 services/core/java/com/android/server/am/ActivityStack.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/am/ActivityStack.java b/services/core/java/com/android/server/am/ActivityStack.java
index 199860597f9..94bef90941c 100644
--- a/services/core/java/com/android/server/am/ActivityStack.java
+++ b/services/core/java/com/android/server/am/ActivityStack.java
@@ -1912,7 +1912,7 @@ class ActivityStack<T extends StackWindowController> extends ConfigurationContai
                         // them. However, when they don't have the wallpaper behind them, we want to
                         // show activities in the next application stack behind them vs. another
                         // task in the home stack like recents.
-                        behindFullscreenActivity = true;
+                        behindFullscreenActivity = task.getTopActivity() != null;
                     } else if (task.isRecentsTask()
                             && task.getTaskToReturnTo() == APPLICATION_ACTIVITY_TYPE) {
                         if (DEBUG_VISIBILITY) Slog.v(TAG_VISIBILITY,
-- 
2.11.0

